### PR TITLE
Supporting importing and exporting API Products

### PR DESCRIPTION
**Pull request by wasuradananjith**
## Purpose
Support importing and exporting API Products from API Controller
Fixes: https://github.com/wso2/product-apim-tooling/issues/291

## Goals
Introduces the functionalities to,
1. Import an API Product
2. Export an API Product

## Approach
1. Introduced the command ```apictl import api-product``` to import an API Product.
image
2. Introduced the command ```apictl export api-product``` to export an API Product.
image
3. After implementing the above functionalities, unit tests have been written for importing and exporting API Products.
4. Upgraded from admin v0.16 REST API to admin v1 REST API.

Refer **[[APIM] Support for API Products from API Controller](http://wso2-oxygen-tank.10903.n7.nabble.com/APIM-Support-for-API-Products-from-API-Controller-td161499.html)** for further information.

## User stories
- Users can export an API Product with the dependent APIs by specifying the mandatory flags and optional flags if necessary. 
- Users can import an API Product with the dependent APIs by specifying the mandatory flags and optional flags if necessary. User can cover the following scenarios in the table by specifying the optional flags **--import-apis**, **--update-api-product** and **--update-apis** as recommended. (To use these commands the user need to have **apim:api_product_import_export** scope. But note that, the scenarios which contain * mark are not allowed to be performed by a user who has only apim:api_product_import_export scope. The user need to have **apim:api_import_export** as well to perform those tasks)
<table>
<tbody>
<tr>
<th>Scenario</th>
<th>--import-apis</th>
<th>--update-api-product</th>
<th>--update-apis</th>
</tr>
<tr>
<td valign="top">
<p>Import a fresh API Product with a fresh set of dependent APIs*.</p>
</td>
<td>
<p>Set (it will be true)</p>
</td>
<td>
<p>Not set (by default false)</p>
</td>
<td>
<p>Not set (by default false)</p>
</td>
</tr>
<tr>
<td valign="top">
<p>Import a fresh API Product when dependent APIs are already imported to APIM successfully and you do not want to update those APIs.</p>
</td>
<td>
<p>Not set (by default false)</p>
</td>
<td>
<p>Not set (by default false)</p>
</td>
<td>
<p>Not set (by default false)</p>
</td>
</tr>
</tr>
<tr>
<td valign="top">
<p>Import a fresh API Product when dependent APIs are already imported to APIM successfully and you want to update those APIs*.</p>
</td>
<td>
<p>You can either set this or not</p>
</td>
<td>
<p>You can either set this or not</p>
</td>
<td>
<p>Set (it will be true)</p>
</td>
</tr>
<tr>
<td valign="top">
<p>Update the API Product only by changing the meta information and by adding/removing the resources of the API Product.</p>
</td>
<td>
<p>Not set (by default false)</p>
</td>
<td>
<p>Set (it will be true)</p>
</td>
<td>
<p>Not set (by default false)</p>
</td>
</tr>
<tr>
<td valign="top">
<p>Update the API Product by adding new resources to both the API Product and the API(s)*.</p>
</td>
<td>
<p>You can either set this or not</p>
</td>
<td>
<p>Not set (by default false)</p>
</td>
<td>
<p>Set (it will be true)</p>
</td>
</tr>
<tr>
<td valign="top">
<p>Update only the dependent APIs*.</p>
</td>
<td>
<p>You can either set this or not</p>
</td>
<td>
<p>Not set (by default false)</p>
</td>
<td>
<p>Set (it will be true)</p>
</td>
</tr>
<tr>
<td valign="top">
<p>Import the API Product and its dependent APIs to another tenant (with --preserve-provider=false) *.</p>
</td>
<td>
<p>Set (it will be true)</p>
</td>
<td>
<p>Not set (by default false)</p>
</td>
<td>
<p>Not set (by default false)</p>
</td>
</tr>
<tr>
<td valign="top">
<p>Update only an already imported API Product and but not its dependent APIs in another tenant (with --preserve-provider=false).</p>
</td>
<td>
<p>Not set (by default false)</p>
</td>
<td>
<p>Set (it will be true)</p>
</td>
<td>
<p>Not set (by default false)</p>
</td>
</tr>
<tr>
<td valign="top">
<p>Update an already imported API Product and its dependent APIs in another tenant (with --preserve-provider=false)*</p>
</td>
<td>
<p>You can either set this or not</p>
</td>
<td>
<p>Not set (by default false)</p>
</td>
<td>
<p>Set (it will be true)</p>
</td>
</tr>

</tbody>
</table>

Simply this means
```
if --import-apis == true {
       // Import the API Product and its dependent APIs
} else If --update-apis == true {
       // Update the dependent APIs  AND the respective API Product
} else if --update-api-products == true {
      // Only update the respective API Product
} else {
     // Only import the API Product
}
```

## Release note
WSO2 API Controller supports importing and exporting API Products now.

## Documentation
Documentation changes should be done. 
https://github.com/wso2/product-apim-tooling/issues/288

## Automation tests
 - Unit tests 
   - product-apim-tooling/import-export-cli/cmd/exportAPIProduct_test.go consists the unit tests for exporting API Products.
   - product-apim-tooling/import-export-cli/cmd/importAPIProduct_test.go consists the unit tests for importing API Products.

## Related PRs
https://github.com/wso2/carbon-apimgt/pull/8500

## Test environment
- Ubuntu 18.04.4 LTS
- go version go1.14 linux/amd64

### CODE REVIEW

File: import-export-cli/cmd/exportAPIProduct.go [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+	}
+}
+
+// WriteAPIProductToZip
+// @param exportAPIProductName : Name of the API Product to be exported
+// @param resp : Response returned from making the HTTP request (only pass a 200 OK)
+// Exported API Product will be written to a zip file
+func WriteAPIProductToZip(exportAPIProductName, exportAPIProductVersion, zipLocationPath string, resp *resty.Response) {
+	// Write to file
+	//directory := filepath.Join(exportDirectory, cmdcmdExportEnvironment)
```
Comment (npamudika) [inline import-export-cli/cmd/exportAPIProduct.go:111]: Let's remove the commented code

Comment (wasuradananjith) [reply import-export-cli/cmd/exportAPIProduct.go:111]: Sure

File: import-export-cli/cmd/exportAPIProduct.go [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+			// If the user has not specified the version, use the version as 1.0.0
+			exportAPIProductVersion = utils.DefaultApiProductVersion
+		}
+		resp, err := getExportApiProductResponse(exportAPIProductName, exportAPIProductVersion, exportAPIProductProvider, exportAPIProductFormat, adminEndpoint,
+			accessToken)
+		if err != nil {
+			utils.HandleErrorAndExit("Error while exporting", err)
+		}
+		// Print info on response
+		utils.Logf(utils.LogPrefixInfo+"ResponseStatus: %v\n", resp.Status())
```
Comment (malinthaprasan) [inline import-export-cli/cmd/exportAPIProduct.go:88]: Shall we add spaces before and after "+".
Saw this in few other places as well, shall we fix all the places?

Comment (malinthaprasan) [inline import-export-cli/cmd/exportAPIProduct.go:88]: Shall we add spaces before and after "+".
Saw this in few other places as well, shall we fix all the places?

**Pull request merged**